src/shared_data.rs

```diff
@@ -0,0 +1,237 @@
+pub enum User {
+    Anyone,
+    Specific(PublicKey),
+}
+
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum Version {
+    FromStart(u64), // Absolute index
+    FromEnd(u64),   // Relative Version - start counting from the end
+}
+
+impl From<u64> for Version {
+    fn from(version: u64) -> Self {
+        Version::FromStart(version)
+    }
+}
+
+// Set of data, owners, permissions versions.
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct ExpectedVersions {
+    expected_data_version: u64,
+    expected_owners_version: u64,
+    expected_access_list_version: u64,
+}
+
+impl ExpectedVersions {
+    pub fn new(
+        expected_data_version: u64,
+        expected_owners_version: u64,
+        expected_access_list_version: u64,
+    ) -> Self {
+        ExpectedVersions {
+            expected_data_version,
+            expected_owners_version,
+            expected_access_list_version,
+        }
+    }
+
+    pub fn expected_data_version(&self) -> u64 {
+        self.expected_data_version
+    }
+
+    pub fn expected_owners_version(&self) -> u64 {
+        self.expected_owners_version
+    }
+
+    pub fn expected_access_list_version(&self) -> u64 {
+        self.expected_access_list_version
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Debug)]
+pub struct Owner {
+    pub public_key: PublicKey,
+    /// The expected Version of the data at the time this ownership change is to become valid.
+    pub expected_data_version: u64,
+    /// The expected Version of the permissions at the time this ownership change is to become valid.
+    pub expected_access_list_version: u64,
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum Kind {
+    PublicSentried,
+    Public,
+    PrivateSentried,
+    Private,
+}
+
+impl Kind {
+    pub fn is_public(self) -> bool {
+        self == Kind::PublicSentried || self == Kind::Public
+    }
+
+    pub fn is_private(self) -> bool {
+        !self.is_public()
+    }
+
+    pub fn is_sentried(self) -> bool {
+        self == Kind::PublicSentried || self == Kind::PrivateSentried
+    }
+
+    /// Creates `Kind` from `public` and `sentried` flags.
+    pub fn from_flags(public: bool, sentried: bool) -> Self {
+        match (public, sentried) {
+            (true, true) => Kind::PublicSentried,
+            (true, false) => Kind::Public,
+            (false, true) => Kind::PrivateSentried,
+            (false, false) => Kind::Private,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum DataAddress {
+    Map(Address),
+    Sequence(Address),
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum Address {
```

Comment:
I'm not entirely sure of this myself. It's also something kept from previous implementation.
I think the end result is that these get different name spaces, so you could store data under same xorname+tag for all four kinds. I'm not totally clear what benefits and drawbacks we have with that.
 Perhaps motivation is found in the RFCs for the Published/Unpublished, Sequenced/Unsequenced data types, otherwise, I think @nbaksalyar and others might have the exact background.